### PR TITLE
Enhance wallet class

### LINKED ISSUE #156: Implement higher level of abstraction such as sendTo(address:), balance()
# What problem does this feture solve?
Currently, receiving and sending Bitcoin are still need long lines of code.

# Describe what you've considered?
Implement higher level of abstraction such as sendTo(address:), balance()

# Code sample / Spec
Wallet class should have such features.

### PR DESCRIPTION
**Pull request by usatie**
### Description of the Change
Implement high level class `Wallet`.
Now everyone can send/receive Bitcoin with only 10 lines.
```
// 1. Create Private Key, Publick Key, Address
let privkey: PrivateKey = PrivateKey(network: .mainnet)
let pubkey: PublicKey = privkey.publicKey()
let address: Cashaddr = pubkey.toCashaddr()
let qrImage: UIImage? = address.qrImage(size: CGSize(width: 240, height: 240))

// 2. Create Wallet, Get balance, Refresh balance
let wallet: Wallet = Wallet(privateKey: privkey,
                            walletDataStore: UserDefaults.standard)
let balance = wallet.balance
wallet.reloadBalance()

// 3. Send Bitcoin
let toAddress = try! AddressFactory.create("bitcoincash:qz936smm809wq0eqthqyh2sgdzhs7v08nvnuq0dmps")
wallet.send(to: toAddress, amount: 10000)
```

#### AddressProvider, UtxoProvider, TransactionHistoryProvider
Basically, both `reload()` and `list()` are methods for retrieving addresses, utxos and transactions. The difference between them is that `reload()` is async method, while `list()` is sync method.

### Alternate Designs
None.

### Benefits
More beginner welcome!

### Possible Drawbacks
None.

### Applicable Issues
closes #156

### CODE REVIEW

File: Sources/BitcoinKit/Wallet/BitcoinCom/BitcoinComService+WalletTransactionProvider.swift [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        return Transaction(version: version, inputs: inputs, outputs: outputs, lockTime: locktime)
+    }
+}
+
+private struct TxIn: Codable {
+    let txid: String
+    let vout: UInt32
+    let sequence: UInt32
+    let scriptSig: ScriptSig
+    // let addr: String
```
Comment (akifuj) [inline Sources/BitcoinKit/Wallet/BitcoinCom/BitcoinComService+WalletTransactionProvider.swift]: Why are they comment out? Please comment the reason somewhere.

File: Sources/BitcoinKit/Wallet/Protocol/AddressProvider.swift [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+public protocol AddressProvider {
+    // GET API: reload utxos
```
Comment (akifuj) [inline Sources/BitcoinKit/Wallet/Protocol/AddressProvider.swift]: This comment is inappropriate, isn't it?

File: Sources/BitcoinKit/Wallet/BitcoinCom/BitcoinComService+WalletTransactionProvider.swift (hunk 1 above)
Comment (usatie) [reply Sources/BitcoinKit/Wallet/BitcoinCom/BitcoinComService+WalletTransactionProvider.swift]: These are part of API response, but not used.

So I commented them out.

File: Sources/BitcoinKit/Wallet/Protocol/AddressProvider.swift (hunk 2 above)
Comment (usatie) [reply Sources/BitcoinKit/Wallet/Protocol/AddressProvider.swift]: Oh it isn't. I'll fix it.

File: Sources/BitcoinKit/Wallet/Protocol/TransactionBuilder.swift [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+public protocol TransactionBuilder {
+    func build(destinations: [(address: Address, amount: UInt64)], utxos: [UnspentTransaction]) throws -> UnsignedTransaction
```
Comment (akifuj) [inline Sources/BitcoinKit/Wallet/Protocol/TransactionBuilder.swift]: The name might be confusing.
 `TransactionBuilder` should build `Transaction` not `UnsignedTransaction`

File: Sources/BitcoinKit/Wallet/Protocol/TransactionProvider.swift [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+public protocol TransactionProvider {
```
Comment (akifuj) [inline Sources/BitcoinKit/Wallet/Protocol/TransactionProvider.swift]: Again, this one also confusing.
How about `TransactionHistoryProvider` .

File: Sources/BitcoinKit/Wallet/Protocol/TransactionProvider.swift [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    // GET API: reload transactions
```
Comment (akifuj) [inline Sources/BitcoinKit/Wallet/Protocol/TransactionProvider.swift]: I think this function would not be only with API, but also SPV or something.

File: Sources/BitcoinKit/Wallet/Protocol/UtxoProvider.swift [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+public protocol UtxoProvider {
+    // GET API: reload utxos
```
Comment (akifuj) [inline Sources/BitcoinKit/Wallet/Protocol/UtxoProvider.swift]: Same as `TransactionProvider` .
Do not need to limit only API.

File: Sources/BitcoinKit/Wallet/Protocol/AddressProvider.swift [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    func reload(keys: [PrivateKey], completion: (([Address]) -> Void)?)
```
Comment (akifuj) [inline Sources/BitcoinKit/Wallet/Protocol/AddressProvider.swift]: I think `convert`  is better as this function name.
Besides, I doubt that this protocol should be under `Wallet` directory.
I think just extension of `PrivateKey` would be better, but still not best idea.

File: Sources/BitcoinKit/Wallet/Protocol/WalletDataStoreProtocol.swift [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+// MARK: - WalletDataStoreProtocol
+public protocol WalletDataStoreProtocol {
+    func getString(forKey key: String) -> String?
+    func setString(_ value: String, forKey key: String)
+}
+
+// MARK: - WalletDataStoreProtocol Extension
+internal enum WalletDataStoreKey: String {
+    case wif, internalIndex, extenralIndex
```
[the remark this answers is not in the document]
Comment (usatie) [reply Sources/BitcoinKit/Wallet/Protocol/WalletDataStoreProtocol.swift]: This protocol will be for HDWallet as well.

File: Sources/BitcoinKit/Wallet/Protocol/AddressProvider.swift (hunk 7 above)
Comment (usatie) [reply Sources/BitcoinKit/Wallet/Protocol/AddressProvider.swift]: No, converting `PrivateKey` to `Address` is not enough. 
Mainly this protocol is for P2SH.

Someone may want to compute new P2SH addresses using the keys outside as well as the keys inside by using API or some other way.

File: Sources/BitcoinKit/Wallet/Protocol/UtxoProvider.swift (hunk 6 above)
Comment (usatie) [reply Sources/BitcoinKit/Wallet/Protocol/UtxoProvider.swift]: Yes

File: Sources/BitcoinKit/Wallet/Protocol/TransactionProvider.swift (hunk 5 above)
Comment (usatie) [reply Sources/BitcoinKit/Wallet/Protocol/TransactionProvider.swift]: Yes

File: Sources/BitcoinKit/Wallet/Protocol/TransactionProvider.swift (hunk 4 above)
Comment (usatie) [reply Sources/BitcoinKit/Wallet/Protocol/TransactionProvider.swift]: This is true. I'll change it.

File: Sources/BitcoinKit/Wallet/Protocol/TransactionBuilder.swift (hunk 3 above)
Comment (usatie) [reply Sources/BitcoinKit/Wallet/Protocol/TransactionBuilder.swift]: I think the name is okay with this. But I think the name of `Transaction` and `UnsignedTransaction` are the problem.

Comment (usatie) [reply Sources/BitcoinKit/Wallet/Protocol/TransactionBuilder.swift]: UnsignedTransaction is almost just a wrap of Transaction and UnspentOutput

File: Sources/BitcoinKit/Wallet/Standard/StandardAddressProvider.swift [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        case cashaddrs
+    }
+
+    public init(userDefaults: UserDefaults) {
+        self.userDefaults = userDefaults
+    }
+
+    public func reload(keys: [PrivateKey], completion: (([Address]) -> Void)?) {
+        let data = try? JSONEncoder().encode(keys.map { $0.publicKey().toCashaddr() })
+        userDefaults.set(data, forKey: UserDefaultsKey.cashaddrs.rawValue)
```
Comment (akifuj) [inline Sources/BitcoinKit/Wallet/Standard/StandardAddressProvider.swift]: You don't need to add `completion?()`?

File: Sources/BitcoinKit/Wallet/Wallet.swift [hunk 10]
```diff
@@ -28,25 +28,144 @@ import Foundation
...
+    public init?(wif: String,
+                 walletDataStore: WalletDataStoreProtocol = UserDefaults.defaultWalletDataStore,
+                 addressProvider: AddressProvider = StandardAddressProvider.shared,
+                 utxoProvider: UtxoProvider = BitcoinComService.shared,
+                 transactionProvider: TransactionHistoryProvider = BitcoinComService.shared,
+                 transactionBroadcaster: TransactionBroadcaster = BitcoinComService.shared,
+                 utxoSelector: UtxoSelector = StandardUtxoSelector.default,
+                 transactionBuilder: TransactionBuilder = StandardTransactionBuilder.default,
+                 transactionSigner: TransactionSigner = StandardTransactionSigner.default) {
+        guard let privkey = try? PrivateKey(wif: wif) else {
```
Comment (akifuj) [inline Sources/BitcoinKit/Wallet/Wallet.swift]: typo

File: Sources/BitcoinKit/Wallet/Wallet.swift [hunk 11]
```diff
@@ -28,25 +28,144 @@ import Foundation
...
+                 utxoProvider: UtxoProvider = BitcoinComService.shared,
+                 transactionProvider: TransactionHistoryProvider = BitcoinComService.shared,
+                 transactionBroadcaster: TransactionBroadcaster = BitcoinComService.shared,
+                 utxoSelector: UtxoSelector = StandardUtxoSelector.default,
+                 transactionBuilder: TransactionBuilder = StandardTransactionBuilder.default,
+                 transactionSigner: TransactionSigner = StandardTransactionSigner.default) {
+        guard let wif = walletDataStore.getString(forKey: .wif) else {
+            return nil
+        }
+        guard let privkey = try? PrivateKey(wif: wif) else {
```
Comment (akifuj) [inline Sources/BitcoinKit/Wallet/Wallet.swift]: typo

File: WalletExample/WalletExample/ViewController.swift [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.createWalletIfNeeded()
+        self.updateLabels()
+    }
+    
+    func createWalletIfNeeded() {
+        if wallet == nil {
+            let privkey = PrivateKey(network: .testnet)
```
Comment (akifuj) [inline WalletExample/WalletExample/ViewController.swift]: typo

File: WalletExample/WalletExample/ViewController.swift [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+            wallet = Wallet(privateKey: privkey)
+            wallet?.save()
+        }
+    }
+    
+    func updateLabels() {
+        qrCodeImageView.image = wallet?.address.qrImage()
+        addressLabel.text = wallet?.address.cashaddr
+        if let balance = wallet?.balance() {
+            balanceLabel.text = "残高：\(balance) satoshi"
```
Comment (akifuj) [inline WalletExample/WalletExample/ViewController.swift]: Use English

File: WalletExample/WalletExample/ViewController.swift [hunk 14]
```diff
@@ -0,0 +1,68 @@
+//
+//  ViewController.swift
+//  WalletExample
+//
+//  Created by Shun Usami on 2018/09/18.
```
Comment (akifuj) [inline WalletExample/WalletExample/ViewController.swift]: Change copyright

File: Sources/BitcoinKit/Wallet/BitcoinCom/BitcoinComEndpoint.swift [hunk 15]
```diff
@@ -24,9 +24,47 @@
 
 import Foundation
 
-public struct BitcoinComEndpoint {
-    public static let testnet: BitcoinComEndpoint = BitcoinComEndpoint(baseUrl: "https://trest.bitcoin.com/v1/")
-    public static let mainnet: BitcoinComEndpoint = BitcoinComEndpoint(baseUrl: "https://rest.bitcoin.com/v1/")
+public struct ApiEndPoint {
```
Comment (akifuj) [inline Sources/BitcoinKit/Wallet/BitcoinCom/BitcoinComEndpoint.swift]: Should you modify file name too?

**Pull request merged**